lld/wasm/Driver.cpp

```diff
@@ -913,7 +913,8 @@ static void createSyntheticSymbols() {
   }
 
   if (ctx.isPic ||
-      config->unresolvedSymbols == UnresolvedPolicy::ImportDynamic) {
+      config->unresolvedSymbols == UnresolvedPolicy::ImportDynamic ||
+      !config->isStatic) {
```

Comment:
I wonder if can replace `ctx.isPic` with ` !config->isStatic` above?    Does it make sense to allow `-shared` + `-static`?  Or `-pie` + `-static`?

I wonder if we need a better way to detect if dynamic linking used at all since you can imagine a case where `isStatic` is true, but we do have dynamic linking.  For example:

```
$ wasm-ld foo.o -Bdynamic -ldylib -Bstatic -lstaticlib
```

In that case since `-Bstatic` comes last on the command line the `config->isStatic` will be false even though we include libdylib.so dyanmically.

Another example:

```
wasm-ld foo.o libdylib.so libstaticlib.a
```

Here `config->isStatic` will remain false even though dynamic linking is used.
```

Reply:
Actually it looks like that the latter case should reject the `.so` file.  I see the ELF linker does this: https://github.com/llvm/llvm-project/blob/96b7c64b8a874584a9dad44bb8901904c14701c0/lld/ELF/Driver.cpp#L341-L344.  I'll upload a PR for that.

Reply:
> I wonder if can replace `ctx.isPic` with ` !config->isStatic` above? Does it make sense to allow `-shared` + `-static`? Or `-pie` + `-static`?
> 
> I wonder if we need a better way to detect if dynamic linking used at all since you can imagine a case where `isStatic` is true, but we do have dynamic linking. For example:
> 
> ```
> $ wasm-ld foo.o -Bdynamic -ldylib -Bstatic -lstaticlib
> ```
> 
> In that case since `-Bstatic` comes last on the command line the `config->isStatic` will be false even though we include libdylib.so dyanmically.

i guess it's simpler to defer the decision to createApplyDataRelocationsFunction.
how do you think?

Reply:
> i guess it's simpler to defer the decision to createApplyDataRelocationsFunction.

i pushed a commit to implement it.

Reply:
@sbc100 do you have any more concerns?